### PR TITLE
Add exponential backoff functions

### PR DESCRIPTION
**Pull request by polyfloyd**

### CODE REVIEW

**Comment by polyfloyd** [comment]
Hmm, there already seems to be a CLA signed for MessageBird. Why didn't the bot detected that?

File: retry/backoff.go [hunk 1]
```diff
@@ -24,3 +24,21 @@ func BackoffLinearWithJitter(waitBetween time.Duration, jitterFraction float64)
 		return backoffutils.JitterUp(waitBetween, jitterFraction)
 	}
 }
+
+// BackoffExponential produces increasing intervals for each attempt.
+//
+// The scalar is multiplied times 2 raised to the current attempt. So the first
+// retry with a scalar of 100ms is 100ms, while the 5th attempt would be 3.2s.
+func BackoffExponential(scalar time.Duration) BackoffFunc {
```
Comment (domgreen) [inline retry/backoff.go]: Please run `make` so that the documentation for this get generated.

Comment (domgreen) [reply retry/backoff.go]: Could you also add an example into `examples_test`

File: retry/backoff.go [hunk 2]
```diff
@@ -24,3 +24,21 @@ func BackoffLinearWithJitter(waitBetween time.Duration, jitterFraction float64)
...
+	return func(attempt uint) time.Duration {
+		return scalar * time.Duration((1<<attempt)>>1)
+	}
+}
+
+// BackoffExponentialWithJitter creates an exponential backoff like
+// BackoffExponential does, but adds jitter.
+func BackoffExponentialWithJitter(scalar time.Duration, jitterFraction float64) BackoffFunc {
```
Comment (domgreen) [inline retry/backoff.go]: As about 👍

File: retry/backoff.go (hunk 1 above)
Comment (polyfloyd) [reply retry/backoff.go]: Done

File: retry/backoff.go [hunk 3]
```diff
@@ -24,3 +24,21 @@ func BackoffLinearWithJitter(waitBetween time.Duration, jitterFraction float64)
...
 	}
 }
+
+// BackoffExponential produces increasing intervals for each attempt.
+//
+// The scalar is multiplied times 2 raised to the current attempt. So the first
+// retry with a scalar of 100ms is 100ms, while the 5th attempt would be 3.2s.
+func BackoffExponential(scalar time.Duration) BackoffFunc {
+	return func(attempt uint) time.Duration {
+		return scalar * time.Duration((1<<attempt)>>1)
```
[the remark this answers is not in the document]
Comment (polyfloyd) [reply retry/backoff.go]: Done

File: retry/backoff.go (hunk 1 above)
Comment (polyfloyd) [reply retry/backoff.go]: Eh, the example bit is done. I ran `make docs`, but it did not produce anything.

Comment (domgreen) [reply retry/backoff.go]: Seems to be some bug in the way that documentation is generated will file an issue. Even if I change `doc.go` it does not correctly change.

**Comment by domgreen** [comment]
Not sure why the CLA is still red could you either ask it to check again or ensure the CLA that has been signed is up to date. 👍

**Comment by arcana261** [comment]
Hi guys!
I wanted the same thing and wanted to contribute but I‌ found someone had already started to do the same awesome job! What's the status of this?‌ I see it's been for a while and CLA is still imposing a problem!

**Comment by polyfloyd** [comment]
`I signed it!`

**Comment by domgreen** [comment]
I'm not sure why the CLA is failing ... any chance you can follow up with on the Google CLA side ... as I cant actually merge without it being signed :cry:

**Comment by domgreen** [comment]
You've probably already checked this but did you sign the SLA with the email your commits are associated with?

**Comment by polyfloyd** [comment]
Actually, I seem to have committed using my personal email address. I'll rewrite the commits with a different address and try again.

**Comment by polyfloyd** [comment]
I signed it!

**Comment by polyfloyd** [comment]
I signed it!

**Comment by polyfloyd** [comment]
Jeez, there we go

**Pull request merged**